src/graph.cpp

```diff
@@ -135,17 +145,67 @@ void Graph::compute_distance_matrix() const {
   }
 }
 
-void Graph::add_node(boost::filesystem::path const &input_file,
+boost::filesystem::path Graph::add_node(boost::filesystem::path const &input_file,
                      Hash const &input_hash) {
-  assert(not has_node(input_file));
-#pragma omp critical(graph_)
-  {
+  if(not has_node(input_file)){
+//#pragma omp critical(graph_)
+  //{
+    if (passed_path_.find(input_file.parent_path()) == passed_path_.end()){ //path not parsed yet
+	passed_path_[input_file.parent_path()] = 1;
+	//std::cout << "YOP\n";
+    }
+
+
+    boost::filesystem::path no_path1 = "/.";
+    boost::filesystem::path no_path2 = ".";
+    std::string file = input_file.string();
+
+    if(input_file.parent_path() == no_path1 || input_file.parent_path() == no_path2){
+	//std::cout<<"/. or .\n";
+
+	for (it_p_=passed_path_.begin(); it_p_!=passed_path_.end();it_p_++){
+		if(it_p_->first != no_path1 && 
+		   it_p_->first != no_path2 &&
+		   has_node(it_p_->first / input_file.filename())){
+			return (it_p_->first / input_file.filename());
```

Comment:
So that's why you changed the add_node return type. Can you explain what's going on?

Reply:
Interesting. I still have a question: even if a library with the same name was already parsed in another directory, why would we link with it? If the library is not in a standard path, there's no reason why we should resolve to that one.

As an exemple, there may be *two* libraries with the same name in two different directories. In that case, to which should we resolve?

Reply:
Once this is settled, you could return a reference to the path stored in the graph instead of making a copy.

Reply:
Actually, if there were (without this) two libraries in different paths, there will be 4 or 5 nodes : 
/lib.dll (not found)
/dir1/lib.dll (first lib parsed)
/dir2/lib.dll (2nd lib parsed)
/home/user/chroot-dir/dir1/lib.dll    or/and    /home/user/chroot-dir/dir2/lib.dll depending the lib in the sub-directories 
 
I had to make a choice, so now there can be 2 different nodes /dir1/lib.dll and /dir2/lib.dll. The most probable reason to have this happening : - scan an entire system so it can mostly be improved if we store the difference 64bits/32bits and link only if it is the same. - chroot folder hand-made and we have stored twice the same dll in different sub-directories.

So, it can be improved, but I think it's better to have 2 nodes if they are in different paths than 4 or 5

Reply:
For the reference, if we don't return (it_p_->first / input_file.filename()), then we have to find mapping_[(it_p_->first / input_file.filename())] then extract the path/filename from his descriptor. because after it use only this. That's a loop (form path/filename to path/filename)we can avoid I think, don't you? (or I didn't get what you meant)